src/tools/r2rdump/R2RDump.cs

```diff
@@ -4,55 +4,436 @@
+                QueryMethod(r2r, "R2R Methods by Query", _queries, true);
+                QueryMethod(r2r, "R2R Methods by Keyword", _keywords, false);
+            }
+
+            Console.WriteLine("========================================================");
+            Console.WriteLine();
+        }
+
+        /// <summary>
+        /// Returns true if the name/signature/id of <param>method</param> matches <param>query</param>
+        /// </summary>
+        /// <param name="exact">Specifies exact or partial match</param>
+        /// <remarks>Case-insensitive and ignores whitespace</remarks>
+        private bool Match(R2RMethod method, string query, bool exact)
+        {
+            bool isNum;
+            int id = ArgStringToInt(query, out isNum);
+            bool idMatch = isNum && (method.Rid == id || method.Token == id);
+
+            bool sigMatch = false;
+            if (exact)
+            {
+                sigMatch = method.Name.Equals(query, StringComparison.OrdinalIgnoreCase);
+                if (!sigMatch)
+                {
+                    string sig = method.Signature.Replace(" ", "");
+                    string q = query.Replace(" ", "");
+                    int iMatch = sig.IndexOf(q, StringComparison.OrdinalIgnoreCase);
+                    sigMatch = (iMatch == 0 || (iMatch > 0 && iMatch == (sig.Length - q.Length) && sig[iMatch - 1] == '.'));
+                }
+            }
+            else
+            {
+                string sig = method.ReturnType + method.Signature.Replace(" ", "");
+                sigMatch = (sig.IndexOf(query.Replace(" ", ""), StringComparison.OrdinalIgnoreCase) >= 0);
+            }
+
+            return idMatch || sigMatch;
+        }
+
+        /// <summary>
+        /// Returns true if the name or value of the ReadyToRunSectionType of <param>section</param> matches <param>query</param>
+        /// </summary>
+        /// <remarks>Case-insensitive</remarks>
+        private bool Match(R2RSection section, string query)
+        {
+            bool isNum;
+            int queryInt = ArgStringToInt(query, out isNum);
+            string typeName = Enum.GetName(typeof(R2RSection.SectionType), section.Type);
+
+            return (isNum && (int)section.Type == queryInt) || typeName.IndexOf(query, StringComparison.OrdinalIgnoreCase) >= 0;
+        }
+
+        /// <summary>
+        /// Finds all R2RMethods by name/signature/id matching <param>query</param>
+        /// </summary>
+        /// <param name="r2r">Contains all extracted info about the ReadyToRun image</param>
+        /// <param name="query">The name/signature/id to search for</param>
+        /// <param name="exact">Specifies exact or partial match</param>
+        /// <out name="res">List of all matching methods</out>
+        /// <remarks>Case-insensitive and ignores whitespace</remarks>
+        public void FindMethod(R2RReader r2r, string query, bool exact, List<R2RMethod> res)
```

Comment:
Instead of adding the R2RMethod to the provided list, can you return the R2RMethod object? While your use case will eventually add it to the list of queried methods, it's possible that future programs could simply ask the R2RReader for a method and then do something else with it.

Reply:
This also applies to FindSection and FindRuntimeFunction

Reply:
If there are more than 1 match for a query, should it return the first one?